src/System.Private.ServiceModel/src/System/IdentityModel/CryptoHelper.cs

```diff
@@ -61,7 +97,305 @@ internal static bool IsAsymmetricAlgorithm(string algorithm)
+            }
+            if (algorithmObject != null)
+            {
+                SymmetricAlgorithm symmetricAlgorithm = algorithmObject as SymmetricAlgorithm;
+                KeyedHashAlgorithm keyedHashAlgorithm = algorithmObject as KeyedHashAlgorithm;
+
+                if (symmetricAlgorithm != null || keyedHashAlgorithm != null)
+                    found = true;
+                // The reason we do not return here even when the user has provided a custom algorithm to CryptoConfig 
+                // is because we need to check if the user has overwritten an existing standard URI.
+            }
+
+            switch (algorithm)
+            {
+                case SecurityAlgorithms.DsaSha1Signature:
+                case SecurityAlgorithms.RsaSha1Signature:
+                case SecurityAlgorithms.RsaSha256Signature:
+                case SecurityAlgorithms.RsaOaepKeyWrap:
+                case SecurityAlgorithms.RsaV15KeyWrap:
+                    return false;
+                case SecurityAlgorithms.HmacSha1Signature:
+                case SecurityAlgorithms.HmacSha256Signature:
+                case SecurityAlgorithms.Psha1KeyDerivation:
+                case SecurityAlgorithms.Psha1KeyDerivationDec2005:
+                    return true;
+                case SecurityAlgorithms.Aes128Encryption:
+                case SecurityAlgorithms.Aes128KeyWrap:
+                    return keySize >= 128 && keySize <= 256;
+                case SecurityAlgorithms.Aes192Encryption:
+                case SecurityAlgorithms.Aes192KeyWrap:
+                    return keySize >= 192 && keySize <= 256;
+                case SecurityAlgorithms.Aes256Encryption:
+                case SecurityAlgorithms.Aes256KeyWrap:
+                    return keySize == 256;
+                case SecurityAlgorithms.TripleDesEncryption:
+                case SecurityAlgorithms.TripleDesKeyWrap:
+                    return keySize == 128 || keySize == 192;
+                default:
+                    if (found)
+                        return true;
+                    return false;
+                    // We do not expect the user to map the uri of an existing standrad algorithm with say key size 128 bit 
+                    // to a custom algorithm with keySize 192 bits. If he does that, we anyways make sure that we return false.
+            }
+        }
+
+        internal static void FillRandomBytes(byte[] buffer)
+        {
+            RandomNumberGenerator.GetBytes(buffer);
+        }
+
+        /// <summary>
+        /// This generates the entropy using random number. This is usually used on the sending 
+        /// side to generate the requestor's entropy.
+        /// </summary>
+        /// <param name="data">The array to fill with cryptographically strong random nonzero bytes.</param>
+        public static void GenerateRandomBytes(byte[] data)
+        {
+            RandomNumberGenerator.GetNonZeroBytes(data);
```

Comment:
Seems weird to want random non-zero.  That indicates that zero is a delimiter, and is introducing a bias on your data.

Reply:
I agree. All this code was copied from .NET Framework and I thought it was unusual, especially as there's a FillRandomBytes method immediately above which includes zeros. I understand this reduces entropy but it was intentional and I don't know why. I will open an issue to track investigating this and change it if possible but for now I'm going to match the behavior of .NET Framework.

---
In reply to: [273995601](https://github.com/dotnet/wcf/pull/3538#discussion_r273995601) [](ancestors = 273995601)